packages/syft/src/syft/service/action/action_service.py

```diff
@@ -751,6 +755,83 @@ def call_method(
 
+    def unwrap_nested_actionobjects(
+        self, context: AuthedServiceContext, data: Any
+    ) -> Any:
+        """recursively unwraps nested action objects"""
+
+        if isinstance(data, list):
+            return [self.unwrap_nested_actionobjects(context, obj) for obj in data]
+        if isinstance(data, dict):
+            return {
+                key: self.unwrap_nested_actionobjects(context, obj)
+                for key, obj in data.items()
+            }
+        if isinstance(data, ActionObject):
+            res = self.get(context=context, uid=data.id)
+            res = res.ok() if res.is_ok() else res.err()
+            if not isinstance(res, ActionObject):
+                return SyftError(message=f"{res}")
+            else:
+                nested_res = res.syft_action_data
+                if isinstance(nested_res, ActionObject):
+                    raise ValueError(
+                        "More than double nesting of ActionObjects is currently not supported"
+                    )
+                return nested_res
+        return data
+
+    def contains_nested_actionobjects(self, data: Any) -> bool:
+        """
+        returns if this is a list/set/dict that contains ActionObjects
+        """
+
+        def unwrap_collection(col: set | dict | list) -> [Any]:  # type: ignore
+            return_values = []
+            if isinstance(col, dict):
+                values = list(col.values()) + list(col.keys())
+            else:
+                values = list(col)
+            for v in values:
+                if isinstance(v, list | dict | set):
+                    return_values += unwrap_collection(v)
+                else:
+                    return_values.append(v)
+            return return_values
+
+        if isinstance(data, list | dict | set):
+            values = unwrap_collection(data)
+            has_action_object = any(isinstance(x, ActionObject) for x in values)
+            return has_action_object
+        elif isinstance(data, ActionObject):
+            return True
+        return False
+
+    def flatten_action_arg(self, context: AuthedServiceContext, arg: UID) -> UID | None:
+        """ "If the argument is a collection (of collections) of ActionObjects,
+        We want to flatten the collection and upload a new ActionObject that contains
+        its values. E.g. [[ActionObject1, ActionObject2],[ActionObject3, ActionObject4]]
+        -> [[value1, value2],[value3, value4]]
+        """
+        res = self.get(context=context, uid=arg)
+        if res.is_err():
+            return arg
+
+        action_object = res.ok()
+        data = action_object.syft_action_data
+
+        if self.contains_nested_actionobjects(data):
+            new_data = self.unwrap_nested_actionobjects(context, data)
+            # Update existing action object with the new flattened data
+            action_object.syft_action_data_cache = new_data
```

Comment:
Refactored this logic to update existing Action Object with new data instead of creating a new one, in order to preserve permissions on the ActionObject.